title: Do Overdraft infocards help users top-up and avoid arrears? (Part II)
author: Helder Silva 
date: 2022-01-26
region: EU
tags: overdraft, infocards, bank products, credit, top-up, arrears, ab test, write-off
link: https://docs.google.com/presentation/d/1hWUr0Pg-neOKaSv1_JjTMSZENBsatGxZtGZmbEu-C9k/
summary:  In order to answer this question, we ran 3 A/B test experiments: Sixty days usage infocard test - displays information that users started using their arranged overdraft, and is meant to be displayed a repeated time 60 days after it was last shown if applicable. Fifty percent usage infocard test - displays information that users have crossed 50% of their arranged overdraft limit. Quarterly top-up infocard test - displays information that users will be charged their overdraft in one week, sent quarterly (since overdraft interest is changed quarterly). We would recommend adopting the Sixty days usage and quarterly top-up infocard infocards moving forward, and the re-running of the fifty percent usage experiment at a higher usage threshold.

<div class="alert alert-block alert-success">
    <H1>Do Overdraft infocards help users top-up and avoid arrears? (Part II)</H1>
</div>

In order to answer the question above, we ran 3 A/B test experiments:
 - <font color='#48AC98'><b>Sixty days usage infocard test</b></font> - Displays information that users started using their arranged overdraft, and is meant to be displayed a repeated time 60 days after it was last shown if applicable
 - <font color='#48AC98'><b>Fifty percent usage infocard test</b></font>  - Displays information that users have crossed 50% of their arranged overdraft limit. In this test we have 2 variants:
    - One infocard that displays the amount of overdraft users
    - One infocard that displays that they are using 50% of their overdraft limit
 - <font color='#48AC98'><b>Quarterly top-up infocard test</b></font> - Displays information that users will be charged their overdraft in one week, sent quarterly (since overdraft interest is changed quarterly)
    
[Here](https://www.figma.com/file/kaE5VHKLfu82RekVXnd8Aa/Overdraft-experiments?node-id=7%3A182) you can find the designs used for each of these infocards.

## Contents
- [Quarterly top-up reminder infocard test results](#section1)
 - [The Sample](#section1.1)
 - [Infocard Interactions](#section1.2)
 - [Top-up Rates](#section1.3)
 - [3 month arrears arrears](#section1.4)
 - [6 month write-offs](#section1.5)


- [Sixty days usage infocard test results](#section2)
 - [3 month arrears](#section2.1)
 - [6 month write-offs](#section2.2)


- [Fifty Percent usage infocard test results](#section3)
 - [3 month arrears](#section3.1)
 - [6 month write-offs](#section3.2)
 
## Experimental Design
For each of the infocards we want to test, users that meet the infocart trigger conditions will be randomly split into 2 groups:
 - One control group that will not receive the infocard
 - One variant group that will receive the infocard
 
This split was automatically handled by plutonium (the overdraft microservice) as soon as the users met the trigger condition.


### Hypothesis

For each of the experiments, we have 2 success measures, so we'll be defining hypothesis for each of these:

<font color='#266678'><b>Top-up rates increase</b></font>
 - **H0:** There are no significant differences in top-up rate within 7 days of meeting the trigger condition
 - **H1:** The top-up rates for the variant group are significantly higher than the control group within 7 days of meeting the trigger condition

<font color='#266678'><b>Arrears rates decrease</b></font>
 - **H0:** There are no significant differences in arrears rates within 3 months of meeting the trigger condition
 - **H1:** The arrears rates for the variant group are significantly lower than the control group within 3 months of meeting the trigger condition
 
<font color='#266678'><b>Write-off rates decrease</b></font>
 - **H0:** There are no significant differences in write-off rates within 6 months of meeting the trigger condition
 - **H1:** The write-off rates for the variant group are significantly lower than the control group within 6 months of meeting the trigger condition

### Experiment Analysis & Timings
In the [part I](https://research.tech26.de/reports/20210525_overdraft_infocard_ab_tests.html) of the analysis of this experiment, run in June 2021 you can find the top-up rates and preliminary arrears results for the sixty days usage infocard and fifty percent usage infocard. For the conclusions of this experiment, we will consider the top-up results from the first part of the results (since there were no visible changes in these since then), and the remaining conclusions will come from this second part of the research. [Here](https://docs.google.com/presentation/d/1hWUr0Pg-neOKaSv1_JjTMSZENBsatGxZtGZmbEu-C9k/edit#slide=id.g10f990ea5de_0_19) you can find a summary of the main results from both experiments combined.

We will be running a Z-Test for each of the experiments and success measures, and assume a significance level of 0.05.

For the sixty day usage and fifty percent usage experiments, they all started at 2021-04-29 and ended at different times, depending on the necessary time to evaluate after the results (for top-ups we kept our first results from June, and needed 3 months for arrears and 6 months for write-offs). Therefore, the top-ups end date is 2021-06-04, the arrears end date is 2021-09-30 and the write-offs end date is 2021-06-30.

For quarterly top-ups, since these are only sent once each quarter, we’re considering results from end of Q2 and end of Q3 for top-ups and arrears, and only end of Q2 for write-offs. 

### Sample split
In the image below we can see the sample split for each of the experiments:
![overdraft_ab_tests_sample_split](overdraft_ab_tests_sample_split.png)

Both variants in the fifty percent usage experiment share the same control group, so even though the control group looks twice as large here, it corresponds to the same proportion of variant users as in the other 2 experiments.


## Key take-aways:

### Sixty Day Infocard Experiment:
 - We have significantly higher top-up rates for the variant group in all tests of the sixty day experiment. This means that for all users that received the infocars, we verified a 1 percentage point increase in top-up rates for users that received the infocard (from 67.8% to 68.8%).
 - However, here we didn’t see any significant decreases in the arrears or write-off rates.
 - **Recommendation**: Given that top-up rates results for this experiment were quite positive, <font color='#48AC98'><b>we would recommend adopting this infocard moving forward.</b></font> 
 

### Fifty Percent Infocard Experiment:
 - Here we didn’t see any significant decreases in the top-up or write-off rates for both variants of this experiment.
 - We see a significant decrease in arrears rates for all users in the percent variant, but not when filtering for users who logged in the app. Based on that, the percent variant seems to perform slightly better than the amount variant.
 - **Recommendation**: Since we didn't find a consistently favorable trend for variant users in this experiment (neither for top-up rates nor arrears rates), we don’t discourage the usage of this infocard, but <font color='#CDA35F'><b>we would recommend re-running this experiment at a higher usage threshold.</b></font> 
 
### Quarterly Top-up Infocard Experiment:
 - For all users that received the infocards, we verified a 0.9 percentage point increase in top-up rates for variant users (from 50.2% to 51.1%).
 - Here we didn’t see any significant decreases in the write-off rates. We see a significant decrease in arrears rates for all users, but not when filtering for users who logged in the app.
 - **Recommendation**: Given that top-up rates results for this experiment were quite positive, <font color='#48AC98'><b>we would recommend adopting this infocard moving forward.</b></font> 

For some of the arrears rates for the comparison of Viewed inforcard (vs. login in control) and Clicked inforcard (vs. login in control), we found some undesirable cases where the variant arrears were significantly higher than control. Given that these variant users are being compared against a different control group (since control users didn't receive an infocard that could be viewed/ clicked), and these arrears results don't seem to materialize into significant write-off differences, we will not be considering these results in the following conclusions.

In [2]:
%%capture
cd/app/

In [3]:
%%capture
!pip install duckdb
!pip install altair
!pip install statsmodels

In [4]:
import pandas as pd
from utils.datalib_database import df_from_sql

import utils.altair_functions as af
import altair as alt
import numpy as np
import scipy.stats as stats
import statsmodels.api as sm

import duckdb

con = duckdb.connect(database=":memory:", read_only=False)

# define colors
color_scale_1 = alt.Scale(domain=["variant", "control"], range=[af.teal, "#98A5A5"])

color_scale_2 = alt.Scale(
    domain=["control", "reminder_use", "reminder_percent"],
    range=["#98A5A5", af.petrol, af.teal],
)


# Build experiment results chart
def experiment_charts(df, title, max_scale, scale):
    all_chart = (
        alt.Chart(df)
        .mark_rect(size=80)
        .encode(
            x=alt.X("user_split:N", title=None),
            y=alt.Y(
                "perc_all_users:Q",
                title="% of users",
                scale=alt.Scale(domain=(0, max_scale)),
            ),
            color=alt.Color(
                "user_split:N",
                scale=scale,
            ),
        )
        .properties(width=200, height=400, title="All users with " + title)
    )

    all_text = all_chart.mark_text(
        align="center",
        baseline="middle",
        # Nudges text to right so it doesn't appear on top of the bar
        dy=-10,
    ).encode(text="perc_all_users:Q")

    login_chart = (
        alt.Chart(df)
        .mark_rect(size=80)
        .encode(
            x=alt.X("user_split:N", title=None),
            y=alt.Y(
                "perc_login_users:Q",
                title="% of users",
                scale=alt.Scale(domain=(0, max_scale)),
            ),
            color=alt.Color(
                "user_split:N",
                scale=color_scale_1,
            ),
        )
        .properties(width=200, height=400, title="Login users with " + title)
    )

    login_text = login_chart.mark_text(
        align="center",
        baseline="middle",
        # Nudges text to right so it doesn't appear on top of the bar
        dy=-10,
    ).encode(text="perc_login_users:Q")

    infocard_view_chart = (
        alt.Chart(df)
        .mark_rect(size=80)
        .encode(
            x=alt.X("user_split:N", title=None),
            y=alt.Y(
                "perc_infocard_view_users:Q",
                title="% of users",
                scale=alt.Scale(domain=(0, max_scale)),
            ),
            color=alt.Color(
                "user_split:N",
                scale=color_scale_1,
            ),
        )
        .properties(width=100, height=400, title="Infocard view users with " + title)
    )

    infocard_view_text = infocard_view_chart.mark_text(
        align="center",
        baseline="middle",
        # Nudges text to right so it doesn't appear on top of the bar
        dy=-10,
    ).encode(text="perc_infocard_view_users:Q")

    infocard_clicked_chart = (
        alt.Chart(df)
        .mark_rect(size=80)
        .encode(
            x=alt.X("user_split:N", title=None),
            y=alt.Y(
                "perc_infocard_click_users:Q",
                title="% of users",
                scale=alt.Scale(domain=(0, max_scale)),
            ),
            color=alt.Color(
                "user_split:N",
                scale=color_scale_1,
            ),
        )
        .properties(width=100, height=400, title="Infocard clicked users with " + title)
    )

    infocard_clicked_text = infocard_clicked_chart.mark_text(
        align="center",
        baseline="middle",
        # Nudges text to right so it doesn't appear on top of the bar
        dy=-10,
    ).encode(text="perc_infocard_click_users:Q")

    return (
        all_chart + all_text
        | login_chart + login_text
        | infocard_view_chart + infocard_view_text
        | infocard_clicked_chart + infocard_clicked_text
    )


# Z-test Function
def z_test(variant_success, variant_obs, control_success, control_obs, alternative):
    p_variant = variant_success / variant_obs
    p_control = control_success / control_obs

    diff = p_variant - p_control

    p_pooled = (variant_success + control_success) * (
        (variant_obs + control_obs) ** (-1)
    )

    n_obs_fact = 1 / variant_obs + 1 / control_obs

    var = (p_pooled * (1 - p_pooled)) * n_obs_fact
    std_diff = var**0.5

    z = diff / std_diff

    if alternative == "two-sided":
        pvalue = stats.norm.sf(abs(z)) * 2
    elif alternative == "larger":
        pvalue = stats.norm.sf(z)
    elif alternative == "smaller":
        pvalue = stats.norm.cdf(z)
    else:
        raise ValueError(
            "Invalid alternative: alternative must be one of two-sided, larger or smaller"
        )

    return round(pvalue, 2)

In [5]:
experiment_query = """ 
with dups as (
    select user_id, 
        count(*) as n_experiments
    from pu_experiments_result
    where created between '2021-04-29' and '2021-09-30'-- excluding dates when card was not visible in dash
group by 1
),
top_ups as (
    select 
        user_id,
        to_char(txn_date, 'YYYY-MM-DD') as top_up_date
    from dbt.zrh_txn_day 
    inner join dbt.zrh_users using (user_created)
    inner join pu_experiments_result pu using (user_id)
    where n_ext_total_in > 0
        and txn_date >= '2021-04-29'
    group by 1, 2
), 
arrears as (
    select 
    user_id,
    to_char(end_time, 'YYYY-MM-DD') as arrears_date
    from dbt.bp_overdraft_users
    inner join dbt.zrh_users using (user_created)
    inner join pu_experiments_result using (user_id)
    where timeframe = 'day'
    and od_enabled_flag 
    and outstanding_balance_eur > max_amount_cents::numeric/100
    and end_time >= '2021-04-29'
    group by 1, 2
),
write_off as (
    select user_id,
    to_char(write_off_dt, 'YYYY-MM-DD') as write_off_date
    from dbt.write_off 
    inner join dbt.zrh_users using (user_created)
    where reason = 'Arranged Overdraft'
    group by 1, 2
)
select 
    to_char(created, 'YYYY-MM-DD') as created, 
    pu.user_id,
    experiment_name, 
    user_in_control_group,
    coalesce(experiment_outcome, 'control') as experiment_outcome,
    n_experiments,
    case when max_amount_cents::numeric = 0 then 0 else round(outstanding_balance_eur::numeric/(max_amount_cents::numeric/100), 2)*100 end as perc_usage,
    max(case when t.user_id is not null then 1 else 0 end) as has_top_up,
    max(case when a.user_id is not null then 1 else 0 end) as has_arrears,
    max(case when wo.user_id is not null then 1 else 0 end) as has_write_off,
    max(case when ld.user_created is not null then 1 else 0 end) as has_login,
    n_experiments - 1 as both_exp_dummy,
    case when user_in_control_group then 1 else 0 end as control_dummy
from pu_experiments_result pu
inner join dbt.zrh_users u using (user_id)
left join dups using (user_id)
left join top_ups t 
    on t.user_id = pu.user_id
    and top_up_date::date between created::date and created::date + interval '7 day'
left join arrears a 
    on a.user_id = pu.user_id
    and arrears_date::date between created::date and created::date + interval '90 day'
left join write_off wo 
    on wo.user_id = pu.user_id
    and write_off_date::date between created::date and created::date + interval '180 day'
left join dbt.zrh_login_day ld
    on ld.user_created = u.user_created
    and login_date between created::date and created::date + interval '7 day'
left join dbt.bp_overdraft_users ou
    on u.user_created = ou.user_created 
    and timeframe = 'day' 
    and end_time::date = created::date
where created between '2021-04-29' and '2021-09-30' -- excluding dates when card was not visible in dash
group by 1, 2, 3, 4, 5, 6, 7
"""

quarterly_infocards_query = """
with infocards as (
    select 
        user_id,
        name as infocard_name,
        count(case when se_label = 'cta1' then 1 end) as n_action_clicked, 
        count(case when se_label = 'cta2' then 1 end) as n_dismissed_clicked, 
        count(case when se_label is null then 1 end) as n_views
    from dbt.stg_txn_events c
    inner join mcv_infocard mi on mi.id = c.se_property
    inner join mcv_infocard_template mit on mi.template_id = mit.id
    where c.event_type in ('686', '687', '454131163','881898916','-14084404','-56984998','-894','6257')
        and c.event_dt between '2021-04-29' and '2021-09-30'
        and mi.created between '2021-04-29' and '2021-09-30'
        and name in ('infocard_overdraft_interest_charge_reminders')
    group by 1, 2
)
select 
    user_id,
    experiment_name, 
    coalesce(experiment_outcome, 'control') as experiment_outcome, 
    user_in_control_group,
    infocard_name,
    coalesce(n_action_clicked, 0) as n_action_clicked, 
    coalesce(n_dismissed_clicked, 0) as n_dismissed_clicked, 
    coalesce(n_views, 0) as n_views
from pu_experiments_result pu 
left join infocards using (user_id)
where created between '2021-04-29' and '2021-09-30'
    and experiment_name = 'quarterly-top-up-reminder'
"""

In [6]:
experiment_df = df_from_sql("redshiftreader", experiment_query)

In [7]:
quarterly_infocards_df = df_from_sql("redshiftreader", quarterly_infocards_query)

<a id='section1'></a>
# Quarterly top-up reminder infocard test results

<a id='section1.1'></a>
## The Sample

Below we can see that we have an evenly split sample of about 36K users for this experiment. Even though the write-off subset of users is smaller than this one, it is equally split as well (as seen in the sample split above).

In [8]:
quarterly_query = """
select * from experiment_df
where experiment_name = 'quarterly-top-up-reminder'
"""

quarterly_sample_query = """
select 
case when user_in_control_group then 'control' else 'variant' end as user_split,
count(*) as n_users
from quarterly_df
group by 1
"""

In [9]:
quarterly_df = con.execute(quarterly_query).fetchdf()

con.register("quarterly_df", quarterly_df)
quarterly_sample_df = con.execute(quarterly_sample_query).fetchdf()

In [10]:
# build chart
chart = (
    alt.Chart(quarterly_sample_df)
    .mark_bar()
    .encode(
        x=alt.X("n_users:Q", title="Customers"),
        y=alt.Y("user_split:N", title=None),
        color=alt.Color(
            "user_split:N",
            scale=color_scale_1,
        ),
    )
    .properties(
        width=600,
        height=200,
        title="Customers in the quarterly top-up reminder infocard experiment",
    )
)

text = chart.mark_text(
    align="center",
    baseline="middle",
    # Nudges text to right so it doesn't appear on top of the bar
    dx=20,
).encode(text="n_users:Q")

chart + text

alt.LayerChart(...)

<a id='section1.2'></a>
## Infocard Interactions

Below you have an overview of how users interact with the N26 app after being selected for this experiment.

In [11]:
quarterly_infocards_query = """
with unions as (
select 
'logins' as event_type,
count(case when not user_in_control_group and has_login = 0 then 1 end) as no_event,
count(case when not user_in_control_group and has_login > 0 then 1 end) as has_event
from quarterly_df
union all 
select 
'views' as event_type,
count(case when not user_in_control_group and n_views = 0 then 1 end) as no_event,
count(case when not user_in_control_group and n_views > 0 then 1 end) as has_event
from quarterly_infocards_df
union all 
select 
'dismissed_clicked' as event_type,
count(case when not user_in_control_group and n_dismissed_clicked = 0 then 1 end) as no_event,
count(case when not user_in_control_group and n_dismissed_clicked > 0 then 1 end) as has_event
from quarterly_infocards_df
union all 
select 
'action_clicked' as event_type,
count(case when not user_in_control_group and n_action_clicked = 0 then 1 end) as no_event,
count(case when not user_in_control_group and n_action_clicked > 0 then 1 end) as has_event
from quarterly_infocards_df
)
select *,
round(has_event::numeric/ (no_event + has_event), 3) * 100 as perc_has_event
from unions
"""

In [12]:
con.register("quarterly_infocards_df", quarterly_infocards_df)
quarterly_infocards_chart_df = con.execute(quarterly_infocards_query).fetchdf()
quarterly_infocards_chart_df

event_type  no_event  has_event  perc_has_event
0             logins     10470      25329            70.8
1              views     13349      22450            62.7
2  dismissed_clicked     24620      11179            31.2
3     action_clicked     33424       2375             6.6

In [13]:
# add login events
af.column_single_label(
    quarterly_infocards_chart_df,
    af.teal,
    "event_type:O",
    "perc_has_event:Q",
    300,
    400,
    "-y",
).properties(
    title="% of events out of all users that received the quarterly top-up reminder infocard"
)

alt.LayerChart(...)

<a id='section1.3'></a>
## Top-up Rates

In [14]:
# Z-Test Function
def quarterly_user_data(col):
    return quarterly_test_df.iloc[0][col].astype(float)


quarterly_test_query = """
select 
--All Sixty Day Users
sum(case when sd.user_in_control_group then 1 end) as quarterly_control_users,
sum(case when not sd.user_in_control_group then 1 end) as quarterly_variant_users,
sum(case when sd.user_in_control_group and has_top_up = 1 then 1 end) as quarterly_control_users_with_top_up,
sum(case when not sd.user_in_control_group and has_top_up = 1 then 1 end) as quarterly_variant_users_with_top_up,
sum(case when sd.user_in_control_group and has_arrears = 1 then 1 end) as quarterly_control_users_with_arrears,
sum(case when not sd.user_in_control_group and has_arrears = 1 then 1 end) as quarterly_variant_users_with_arrears,
sum(case when sd.user_in_control_group and has_write_off = 1 then 1 end) as quarterly_control_users_with_write_off,
sum(case when not sd.user_in_control_group and has_write_off = 1 then 1 end) as quarterly_variant_users_with_write_off,

-- Sixty Day Users With Login
sum(case when sd.user_in_control_group and has_login = 1 then 1 end) as quarterly_control_users_with_login,
sum(case when not sd.user_in_control_group and has_login = 1 then 1 end) as quarterly_variant_users_with_login,
sum(case when sd.user_in_control_group and has_login = 1 and has_top_up = 1 then 1 end) as quarterly_control_users_with_top_up_and_login,
sum(case when not sd.user_in_control_group and has_login = 1 and has_top_up = 1 then 1 end) as quarterly_variant_users_with_top_up_and_login,
sum(case when sd.user_in_control_group and has_login = 1 and has_arrears = 1 then 1 end) as quarterly_control_users_with_arrears_and_login,
sum(case when not sd.user_in_control_group and has_login = 1 and has_arrears = 1 then 1 end) as quarterly_variant_users_with_arrears_and_login,
sum(case when sd.user_in_control_group and has_login = 1 and has_write_off = 1 then 1 end) as quarterly_control_users_with_write_off_and_login,
sum(case when not sd.user_in_control_group and has_login = 1 and has_write_off = 1 then 1 end) as quarterly_variant_users_with_write_off_and_login,

-- Sixty Day Users With Infocard Views
sum(case when not sd.user_in_control_group and n_views >0 then 1 end) as quarterly_variant_users_with_view,
sum(case when not sd.user_in_control_group and n_views >0 and has_top_up = 1 then 1 end) as quarterly_variant_users_with_top_up_and_view,
sum(case when not sd.user_in_control_group and n_views >0 then 1 end) as quarterly_variant_users_with_view,
sum(case when not sd.user_in_control_group and n_views >0 and has_arrears = 1 then 1 end) as quarterly_variant_users_with_arrears_and_view, 
sum(case when not sd.user_in_control_group and n_views >0 and has_write_off = 1 then 1 end) as quarterly_variant_users_with_write_off_and_view, 
-- Sixty Day Users With Infocard Clicks
sum(case when not sd.user_in_control_group and n_action_clicked >0 then 1 end) as quarterly_variant_users_with_click,
sum(case when not sd.user_in_control_group and n_action_clicked >0 and has_top_up = 1 then 1 end) as quarterly_variant_users_with_top_up_and_click,
sum(case when not sd.user_in_control_group and n_action_clicked >0 then 1 end) as quarterly_variant_users_with_click,
sum(case when not sd.user_in_control_group and n_action_clicked >0 and has_arrears = 1 then 1 end) as quarterly_variant_users_with_arrears_and_click,
sum(case when not sd.user_in_control_group and n_action_clicked >0 and has_write_off = 1 then 1 end) as quarterly_variant_users_with_write_off_and_click
from quarterly_df sd
left join quarterly_infocards_df using (user_id)
order by 2, 1
"""
quarterly_test_df = con.execute(quarterly_test_query).fetchdf()

In [15]:
quarterly_top_up_query = """ 
with totals as (
select 
case when pu.user_in_control_group then 'control' else 'variant' end as user_split,
case when has_top_up = 1 then 'Has Top-up' else 'No Top-up' end as top_up_type,
count(*) as n_users,
count(case when has_login = 1 then 1 end) as login_users, 
count(case when n_views>0 then 1 end) as infocard_view_users,
count(case when n_action_clicked>0 then 1 end) as infocard_clicked_users
from quarterly_df pu
left join quarterly_infocards_df using (user_id)
group by 1, 2
)
select *, 
round(n_users::numeric/ (sum(n_users) over (partition by user_split))::numeric, 3)*100 as perc_all_users,
round(login_users::numeric/ (sum(login_users) over (partition by user_split))::numeric, 3)*100 as perc_login_users,
round(infocard_view_users::numeric/ (sum(infocard_view_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_view_users,
round(infocard_clicked_users::numeric/ (sum(infocard_clicked_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_click_users
from totals
order by 2, 1
"""

In [16]:
quarterly_top_up_df = con.execute(quarterly_top_up_query).fetchdf()
quarterly_top_up_df

user_split top_up_type  n_users  login_users  infocard_view_users  \
0    control  Has Top-up    18177        17340                    0   
1    variant  Has Top-up    18287        17477                14884   
2    control   No Top-up    18001         8179                    0   
3    variant   No Top-up    17512         7852                 7566   

   infocard_clicked_users  perc_all_users  perc_login_users  \
0                       0            50.2              67.9   
1                    1741            51.1              69.0   
2                       0            49.8              32.1   
3                     634            48.9              31.0   

   perc_infocard_view_users  perc_infocard_click_users  
0                       NaN                        NaN  
1                      66.3                       73.3  
2                       NaN                        NaN  
3                      33.7                       26.7

In [17]:
experiment_charts(
    quarterly_top_up_df[quarterly_top_up_df["top_up_type"] == "Has Top-up"],
    "top-up",
    100,
    color_scale_1,
)

alt.HConcatChart(...)

#### Top-up Z-Test results for the Sixty Day Reminder Infocard

In [18]:
pd.set_option("display.max_colwidth", None)

In [19]:
top_up_sixty_all_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_top_up"),
    quarterly_user_data("quarterly_variant_users"),
    quarterly_user_data("quarterly_control_users_with_top_up"),
    quarterly_user_data("quarterly_control_users"),
    "larger",
)
top_up_sixty_login_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_top_up_and_login"),
    quarterly_user_data("quarterly_variant_users_with_login"),
    quarterly_user_data("quarterly_control_users_with_top_up_and_login"),
    quarterly_user_data("quarterly_control_users_with_login"),
    "larger",
)
top_up_sixty_view_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_top_up_and_view"),
    quarterly_user_data("quarterly_variant_users_with_view"),
    quarterly_user_data("quarterly_control_users_with_top_up_and_login"),
    quarterly_user_data("quarterly_control_users_with_login"),
    "larger",
)
top_up_sixty_click_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_top_up_and_click"),
    quarterly_user_data("quarterly_variant_users_with_click"),
    quarterly_user_data("quarterly_control_users_with_top_up_and_login"),
    quarterly_user_data("quarterly_control_users_with_login"),
    "larger",
)

p_value_top_up_quarterly_df = pd.DataFrame(
    {
        "test": [
            "All users with top-up",
            "Users that logged-in",
            "Viewed inforcard (vs. login in control)",
            "Clicked inforcard (vs. login in control)",
        ],
        "larger_p_value": [
            top_up_sixty_all_larger,
            top_up_sixty_login_larger,
            top_up_sixty_view_larger,
            top_up_sixty_click_larger,
        ],
    }
)

p_value_top_up_quarterly_df["smaller_p_value"] = (
    1 - p_value_top_up_quarterly_df["larger_p_value"]
)
p_value_top_up_quarterly_df["result"] = np.where(
    p_value_top_up_quarterly_df["larger_p_value"] < 0.05,
    "Variant top-up rates are significantly higher than the control (reject H0)",
    "Samples have similar top-up rates (fail to reject H0)",
)

p_value_top_up_quarterly_df

test  larger_p_value  smaller_p_value  \
0                     All users with top-up            0.01             0.99   
1                      Users that logged-in            0.01             0.99   
2   Viewed inforcard (vs. login in control)            1.00             0.00   
3  Clicked inforcard (vs. login in control)            0.00             1.00   

                                                                       result  
0  Variant top-up rates are significantly higher than the control (reject H0)  
1  Variant top-up rates are significantly higher than the control (reject H0)  
2                       Samples have similar top-up rates (fail to reject H0)  
3  Variant top-up rates are significantly higher than the control (reject H0)

**Conclusion:** We have significantly higher top-up rates for the variant group in almost all tests of the quarterly top-up experiment (except for users who viewed the infocard). This means that for all users that received the infocards, we verified a 0.9 percentage point increase in top-up rates for variant users (from 50.2% to 51.1%)

<a id='section1.4'></a>
## 3 month arrears

In [20]:
quarterly_arrears_query = """
with totals as (
select 
case when pu.user_in_control_group then 'control' else 'variant' end as user_split,
case when has_arrears = 1 then 'Has Arrears' else 'No Arrears' end as arrears_type,
count(*) as n_users,
count(case when has_login = 1 then 1 end) as login_users, 
count(case when n_views>0 then 1 end) as infocard_view_users,
count(case when n_action_clicked>0 then 1 end) as infocard_clicked_users
from quarterly_df pu
left join quarterly_infocards_df using (user_id)
group by 1, 2
)
select *, 
round(n_users::numeric/ (sum(n_users) over (partition by user_split))::numeric, 3)*100 as perc_all_users,
round(login_users::numeric/ (sum(login_users) over (partition by user_split))::numeric, 3)*100 as perc_login_users,
round(infocard_view_users::numeric/ (sum(infocard_view_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_view_users,
round(infocard_clicked_users::numeric/ (sum(infocard_clicked_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_click_users
from totals
order by 2, 1
"""

In [21]:
quarterly_arrears_df = con.execute(quarterly_arrears_query).fetchdf()
quarterly_arrears_df

user_split arrears_type  n_users  login_users  infocard_view_users  \
0    control  Has Arrears     5669         4119                    0   
1    variant  Has Arrears     5438         3971                 3849   
2    control   No Arrears    30509        21400                    0   
3    variant   No Arrears    30361        21358                18601   

   infocard_clicked_users  perc_all_users  perc_login_users  \
0                       0            15.7              16.1   
1                     374            15.2              15.7   
2                       0            84.3              83.9   
3                    2001            84.8              84.3   

   perc_infocard_view_users  perc_infocard_click_users  
0                       NaN                        NaN  
1                      17.1                       15.7  
2                       NaN                        NaN  
3                      82.9                       84.3

In [22]:
experiment_charts(
    quarterly_arrears_df[quarterly_arrears_df["arrears_type"] == "Has Arrears"],
    "arrears",
    20,
    color_scale_1,
)

alt.HConcatChart(...)

#### Arrears Z-Test results for the Sixty Day Reminder Infocard

In [23]:
arrears_sixty_all_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_arrears"),
    quarterly_user_data("quarterly_variant_users"),
    quarterly_user_data("quarterly_control_users_with_arrears"),
    quarterly_user_data("quarterly_control_users"),
    "larger",
)
arrears_sixty_login_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_arrears_and_login"),
    quarterly_user_data("quarterly_variant_users_with_login"),
    quarterly_user_data("quarterly_control_users_with_arrears_and_login"),
    quarterly_user_data("quarterly_control_users_with_login"),
    "larger",
)
arrears_sixty_view_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_arrears_and_view"),
    quarterly_user_data("quarterly_variant_users_with_view"),
    quarterly_user_data("quarterly_control_users_with_arrears_and_login"),
    quarterly_user_data("quarterly_control_users_with_login"),
    "larger",
)
arrears_sixty_click_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_arrears_and_click"),
    quarterly_user_data("quarterly_variant_users_with_click"),
    quarterly_user_data("quarterly_control_users_with_arrears_and_login"),
    quarterly_user_data("quarterly_control_users_with_login"),
    "larger",
)

p_value_arrears_quarterly_df = pd.DataFrame(
    {
        "test": [
            "All users with top-up",
            "Users that logged-in",
            "Viewed inforcard (vs. login in control)",
            "Clicked inforcard (vs. login in control)",
        ],
        "larger_p_value": [
            arrears_sixty_all_larger,
            arrears_sixty_login_larger,
            arrears_sixty_view_larger,
            arrears_sixty_click_larger,
        ],
    }
)

p_value_arrears_quarterly_df["smaller_p_value"] = (
    1 - p_value_arrears_quarterly_df["larger_p_value"]
)
p_value_arrears_quarterly_df["result"] = np.where(
    p_value_arrears_quarterly_df["smaller_p_value"] <= 0.05,
    "Variant arrears rates are significantly lower than the control (reject H0)",
    np.where(
        p_value_arrears_quarterly_df["smaller_p_value"] >= 0.95,
        "Variant arrears rates are significantly higher the control (reject H0 but undesirable effect)",
        "Samples have similar arrears rates (fail to reject H0)",
    ),
)
p_value_arrears_quarterly_df

test  larger_p_value  smaller_p_value  \
0                     All users with top-up            0.96             0.04   
1                      Users that logged-in            0.92             0.08   
2   Viewed inforcard (vs. login in control)            0.00             1.00   
3  Clicked inforcard (vs. login in control)            0.69             0.31   

                                                                                          result  
0                     Variant arrears rates are significantly lower than the control (reject H0)  
1                                         Samples have similar arrears rates (fail to reject H0)  
2  Variant arrears rates are significantly higher the control (reject H0 but undesirable effect)  
3                                         Samples have similar arrears rates (fail to reject H0)

**Conclusion:** With a decrease of 0.4 percentage points from 15.7% to 15.2%, we have significant lower arrears rates for the variant group in the quarterly top-up experiment for all users. Also, for the variant users who viewed the infocard vs. login users in control, we seem to have an undesirable significant increase of 1 percentage point.

<a id='section1.5'></a>
## 6 month write-offs

In [24]:
quarterly_write_off_query = """
with totals as (
select 
case when pu.user_in_control_group then 'control' else 'variant' end as user_split,
case when has_write_off = 1 then 'Has Write-off' else 'No Write-off' end as write_off_type,
count(*) as n_users,
count(case when has_login = 1 then 1 end) as login_users, 
count(case when n_views>0 then 1 end) as infocard_view_users,
count(case when n_action_clicked>0 then 1 end) as infocard_clicked_users
from quarterly_df pu
left join quarterly_infocards_df using (user_id)
group by 1, 2
)
select *, 
round(n_users::numeric/ (sum(n_users) over (partition by user_split))::numeric, 3)*100 as perc_all_users,
round(login_users::numeric/ (sum(login_users) over (partition by user_split))::numeric, 3)*100 as perc_login_users,
round(infocard_view_users::numeric/ (sum(infocard_view_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_view_users,
round(infocard_clicked_users::numeric/ (sum(infocard_clicked_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_click_users
from totals
order by 2, 1
"""

In [25]:
quarterly_write_off_df = con.execute(quarterly_write_off_query).fetchdf()
quarterly_write_off_df

user_split write_off_type  n_users  login_users  infocard_view_users  \
0    control  Has Write-off      676          324                    0   
1    variant  Has Write-off      662          330                  254   
2    control   No Write-off    35502        25195                    0   
3    variant   No Write-off    35137        24999                22196   

   infocard_clicked_users  perc_all_users  perc_login_users  \
0                       0             1.9               1.3   
1                      31             1.8               1.3   
2                       0            98.1              98.7   
3                    2344            98.2              98.7   

   perc_infocard_view_users  perc_infocard_click_users  
0                       NaN                        NaN  
1                       1.1                        1.3  
2                       NaN                        NaN  
3                      98.9                       98.7

In [26]:
experiment_charts(
    quarterly_write_off_df[quarterly_write_off_df["write_off_type"] == "Has Write-off"],
    "write-offs",
    3,
    color_scale_1,
)

alt.HConcatChart(...)

In [27]:
write_off_sixty_all_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_write_off"),
    quarterly_user_data("quarterly_variant_users"),
    quarterly_user_data("quarterly_control_users_with_write_off"),
    quarterly_user_data("quarterly_control_users"),
    "larger",
)
write_off_sixty_login_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_write_off_and_login"),
    quarterly_user_data("quarterly_variant_users_with_login"),
    quarterly_user_data("quarterly_control_users_with_write_off_and_login"),
    quarterly_user_data("quarterly_control_users_with_login"),
    "larger",
)
write_off_sixty_view_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_write_off_and_view"),
    quarterly_user_data("quarterly_variant_users_with_view"),
    quarterly_user_data("quarterly_control_users_with_write_off_and_login"),
    quarterly_user_data("quarterly_control_users_with_login"),
    "larger",
)
write_off_sixty_click_larger = z_test(
    quarterly_user_data("quarterly_variant_users_with_write_off_and_click"),
    quarterly_user_data("quarterly_variant_users_with_click"),
    quarterly_user_data("quarterly_control_users_with_write_off_and_login"),
    quarterly_user_data("quarterly_control_users_with_login"),
    "larger",
)

p_value_write_off_quarterly_df = pd.DataFrame(
    {
        "test": [
            "All users with top-up",
            "Users that logged-in",
            "Viewed inforcard (vs. login in control)",
            "Clicked inforcard (vs. login in control)",
        ],
        "larger_p_value": [
            write_off_sixty_all_larger,
            write_off_sixty_login_larger,
            write_off_sixty_view_larger,
            write_off_sixty_click_larger,
        ],
    }
)

p_value_write_off_quarterly_df["smaller_p_value"] = (
    1 - p_value_write_off_quarterly_df["larger_p_value"]
)
p_value_write_off_quarterly_df["result"] = np.where(
    p_value_write_off_quarterly_df["smaller_p_value"] <= 0.05,
    "Variant write_off rates are significantly lower than the control (reject H0)",
    np.where(
        p_value_write_off_quarterly_df["smaller_p_value"] >= 0.95,
        "Variant write_off rates are significantly higher the control (reject H0 but undesirable effect)",
        "Samples have similar write_off rates (fail to reject H0)",
    ),
)
p_value_write_off_quarterly_df

test  larger_p_value  smaller_p_value  \
0                     All users with top-up            0.58             0.42   
1                      Users that logged-in            0.37             0.63   
2   Viewed inforcard (vs. login in control)            0.92             0.08   
3  Clicked inforcard (vs. login in control)            0.44             0.56   

                                                     result  
0  Samples have similar write_off rates (fail to reject H0)  
1  Samples have similar write_off rates (fail to reject H0)  
2  Samples have similar write_off rates (fail to reject H0)  
3  Samples have similar write_off rates (fail to reject H0)

**Conclusion:** Here we have no significant decrease in the write-off rates.

<a id='section2'></a>
# Sixty days usage infocard test results

<a id='section2.1'></a>
## 3 month arrears

In [28]:
sixty_day_infocards_query = """
with infocards as (
select 
user_id,
name as infocard_name,
count(case when se_label = 'cta1' then 1 end) as n_action_clicked, 
count(case when se_label = 'cta2' then 1 end) as n_dismissed_clicked, 
count(case when se_label is null then 1 end) as n_views
from dbt.stg_txn_events c
inner join mcv_infocard mi on mi.id = c.se_property
inner join mcv_infocard_template mit on mi.template_id = mit.id
where c.event_type in ('686', '687', '454131163','881898916','-14084404','-56984998','-894','6257')
and c.event_dt between '2021-04-29' and '2021-09-30'
and mi.created between '2021-04-29' and '2021-09-30'
and name in ('infocard_overdraft_first_sixty_day')
group by 1, 2
)
select 
user_id,
experiment_name, 
coalesce(experiment_outcome, 'control') as experiment_outcome, 
user_in_control_group,
infocard_name,
coalesce(n_action_clicked, 0) as n_action_clicked, 
coalesce(n_dismissed_clicked, 0) as n_dismissed_clicked, 
coalesce(n_views, 0) as n_views
from pu_experiments_result pu 
left join infocards
using (user_id)
where created between '2021-04-29' and '2021-09-30'
and experiment_name = 'sixty-days-usage-info-card'
"""

fifty_perc_infocards_query = """
with infocards as (
select 
user_id,
name as infocard_name,
count(case when se_label = 'cta1' then 1 end) as n_action_clicked, 
count(case when se_label = 'cta2' then 1 end) as n_dismissed_clicked, 
count(case when se_label is null then 1 end) as n_views
from dbt.stg_txn_events c
inner join mcv_infocard mi on mi.id = c.se_property
inner join mcv_infocard_template mit on mi.template_id = mit.id
where c.event_type in ('686', '687', '454131163','881898916','-14084404','-56984998','-894','6257')
and c.event_dt between '2021-04-29' and '2021-09-30'
and mi.created between '2021-04-29' and '2021-09-30'
and name in ('infocard_overdraft_limit_reminder_percent', 'infocard_overdraft_limit_reminder_use')
group by 1, 2
)
select 
user_id,
experiment_name, 
coalesce(experiment_outcome, 'control') as experiment_outcome, 
user_in_control_group,
infocard_name,
coalesce(n_action_clicked, 0) as n_action_clicked, 
coalesce(n_dismissed_clicked, 0) as n_dismissed_clicked, 
coalesce(n_views, 0) as n_views
from pu_experiments_result pu 
left join infocards
using (user_id)
where created between '2021-04-29' and '2021-09-30'
and experiment_name = 'fifty-percent-usage-exceeded'
"""

In [29]:
sixty_day_infocards_df = df_from_sql("redshiftreader", sixty_day_infocards_query)

In [30]:
fifty_perc_infocards_df = df_from_sql("redshiftreader", fifty_perc_infocards_query)

In [31]:
sixty_day_query = """
select * from experiment_df
where experiment_name = 'sixty-days-usage-info-card'
"""

sixty_day_sample_query = """
select 
case when user_in_control_group then 'control' else 'variant' end as user_split,
count(*) as n_users
from sixty_day_df
group by 1
"""

In [32]:
sixty_day_df = con.execute(sixty_day_query).fetchdf()

con.register("sixty_day_df", sixty_day_df)
sixty_day_sample_df = con.execute(sixty_day_sample_query).fetchdf()

In [33]:
# Z-Test Function
def sixty_day_user_data(col):
    return sixty_day_test_df.iloc[0][col].astype(float)


sixty_day_test_query = """
select 
--All Sixty Day Users
sum(case when sd.user_in_control_group then 1 end) as sixty_day_control_users,
sum(case when not sd.user_in_control_group then 1 end) as sixty_day_variant_users,
sum(case when sd.user_in_control_group and has_top_up = 1 then 1 end) as sixty_day_control_users_with_top_up,
sum(case when not sd.user_in_control_group and has_top_up = 1 then 1 end) as sixty_day_variant_users_with_top_up,
sum(case when sd.user_in_control_group and has_arrears = 1 then 1 end) as sixty_day_control_users_with_arrears,
sum(case when not sd.user_in_control_group and has_arrears = 1 then 1 end) as sixty_day_variant_users_with_arrears,
sum(case when sd.user_in_control_group and has_write_off = 1 then 1 end) as sixty_day_control_users_with_write_off,
sum(case when not sd.user_in_control_group and has_write_off = 1 then 1 end) as sixty_day_variant_users_with_write_off,
-- Sixty Day Users With Login
sum(case when sd.user_in_control_group and has_login = 1 then 1 end) as sixty_day_control_users_with_login,
sum(case when not sd.user_in_control_group and has_login = 1 then 1 end) as sixty_day_variant_users_with_login,
sum(case when sd.user_in_control_group and has_login = 1 and has_top_up = 1 then 1 end) as sixty_day_control_users_with_top_up_and_login,
sum(case when not sd.user_in_control_group and has_login = 1 and has_top_up = 1 then 1 end) as sixty_day_variant_users_with_top_up_and_login,
sum(case when sd.user_in_control_group and has_login = 1 and has_arrears = 1 then 1 end) as sixty_day_control_users_with_arrears_and_login,
sum(case when not sd.user_in_control_group and has_login = 1 and has_arrears = 1 then 1 end) as sixty_day_variant_users_with_arrears_and_login,
sum(case when sd.user_in_control_group and has_login = 1 and has_write_off = 1 then 1 end) as sixty_day_control_users_with_write_off_and_login,
sum(case when not sd.user_in_control_group and has_login = 1 and has_write_off = 1 then 1 end) as sixty_day_variant_users_with_write_off_and_login,
-- Sixty Day Users With Infocard Views
sum(case when not sd.user_in_control_group and n_views >0 then 1 end) as sixty_day_variant_users_with_view,
sum(case when not sd.user_in_control_group and n_views >0 and has_top_up = 1 then 1 end) as sixty_day_variant_users_with_top_up_and_view,
sum(case when not sd.user_in_control_group and n_views >0 then 1 end) as sixty_day_variant_users_with_view,
sum(case when not sd.user_in_control_group and n_views >0 and has_arrears = 1 then 1 end) as sixty_day_variant_users_with_arrears_and_view, 
sum(case when not sd.user_in_control_group and n_views >0 and has_write_off = 1 then 1 end) as sixty_day_variant_users_with_write_off_and_view, 
-- Sixty Day Users With Infocard Clicks
sum(case when not sd.user_in_control_group and n_action_clicked >0 then 1 end) as sixty_day_variant_users_with_click,
sum(case when not sd.user_in_control_group and n_action_clicked >0 and has_top_up = 1 then 1 end) as sixty_day_variant_users_with_top_up_and_click,
sum(case when not sd.user_in_control_group and n_action_clicked >0 then 1 end) as sixty_day_variant_users_with_click,
sum(case when not sd.user_in_control_group and n_action_clicked >0 and has_arrears = 1 then 1 end) as sixty_day_variant_users_with_arrears_and_click,
sum(case when not sd.user_in_control_group and n_action_clicked >0 and has_write_off = 1 then 1 end) as sixty_day_variant_users_with_write_off_and_click
from sixty_day_df sd
left join sixty_day_infocards_df using (user_id)
order by 2, 1
"""
sixty_day_test_df = con.execute(sixty_day_test_query).fetchdf()

In [34]:
sixty_day_arrears_query = """
with totals as (
select 
case when pu.user_in_control_group then 'control' else 'variant' end as user_split,
case when has_arrears = 1 then 'Has Arrears' else 'No Arrears' end as arrears_type,
count(*) as n_users,
count(case when has_login = 1 then 1 end) as login_users, 
count(case when n_views>0 then 1 end) as infocard_view_users,
count(case when n_action_clicked>0 then 1 end) as infocard_clicked_users
from sixty_day_df pu
left join sixty_day_infocards_df using (user_id)
group by 1, 2
)
select *, 
round(n_users::numeric/ (sum(n_users) over (partition by user_split))::numeric, 3)*100 as perc_all_users,
round(login_users::numeric/ (sum(login_users) over (partition by user_split))::numeric, 3)*100 as perc_login_users,
round(infocard_view_users::numeric/ (sum(infocard_view_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_view_users,
round(infocard_clicked_users::numeric/ (sum(infocard_clicked_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_click_users
from totals
order by 2, 1
"""

In [35]:
sixty_day_arrears_df = con.execute(sixty_day_arrears_query).fetchdf()
sixty_day_arrears_df

user_split arrears_type  n_users  login_users  infocard_view_users  \
0    control  Has Arrears     1766         1718                    0   
1    variant  Has Arrears     1743         1700                 1377   
2    control   No Arrears    30023        27724                    0   
3    variant   No Arrears    30105        27821                21189   

   infocard_clicked_users  perc_all_users  perc_login_users  \
0                       0             5.6               5.8   
1                     175             5.5               5.8   
2                       0            94.4              94.2   
3                    2390            94.5              94.2   

   perc_infocard_view_users  perc_infocard_click_users  
0                       NaN                        NaN  
1                       6.1                        6.8  
2                       NaN                        NaN  
3                      93.9                       93.2

In [36]:
experiment_charts(
    sixty_day_arrears_df[sixty_day_arrears_df["arrears_type"] == "Has Arrears"],
    "arrears",
    10,
    color_scale_1,
)

alt.HConcatChart(...)

In [37]:
arrears_sixty_all_larger = z_test(
    sixty_day_user_data("sixty_day_variant_users_with_arrears"),
    sixty_day_user_data("sixty_day_variant_users"),
    sixty_day_user_data("sixty_day_control_users_with_arrears"),
    sixty_day_user_data("sixty_day_control_users"),
    "larger",
)
arrears_sixty_login_larger = z_test(
    sixty_day_user_data("sixty_day_variant_users_with_arrears_and_login"),
    sixty_day_user_data("sixty_day_variant_users_with_login"),
    sixty_day_user_data("sixty_day_control_users_with_arrears_and_login"),
    sixty_day_user_data("sixty_day_control_users_with_login"),
    "larger",
)
arrears_sixty_view_larger = z_test(
    sixty_day_user_data("sixty_day_variant_users_with_arrears_and_view"),
    sixty_day_user_data("sixty_day_variant_users_with_view"),
    sixty_day_user_data("sixty_day_control_users_with_arrears_and_login"),
    sixty_day_user_data("sixty_day_control_users_with_login"),
    "larger",
)
arrears_sixty_click_larger = z_test(
    sixty_day_user_data("sixty_day_variant_users_with_arrears_and_click"),
    sixty_day_user_data("sixty_day_variant_users_with_click"),
    sixty_day_user_data("sixty_day_control_users_with_arrears_and_login"),
    sixty_day_user_data("sixty_day_control_users_with_login"),
    "larger",
)

p_value_arrears_sixty_days_df = pd.DataFrame(
    {
        "test": [
            "All users with top-up",
            "Users that logged-in",
            "Viewed inforcard (vs. login in control)",
            "Clicked inforcard (vs. login in control)",
        ],
        "larger_p_value": [
            arrears_sixty_all_larger,
            arrears_sixty_login_larger,
            arrears_sixty_view_larger,
            arrears_sixty_click_larger,
        ],
    }
)

p_value_arrears_sixty_days_df["smaller_p_value"] = (
    1 - p_value_arrears_sixty_days_df["larger_p_value"]
)
p_value_arrears_sixty_days_df["result"] = np.where(
    p_value_arrears_sixty_days_df["smaller_p_value"] <= 0.05,
    "Variant arrears rates are significantly lower than the control (reject H0)",
    np.where(
        p_value_arrears_sixty_days_df["smaller_p_value"] >= 0.95,
        "Variant arrears rates are significantly higher the control (reject H0 but undesirable effect)",
        "Samples have similar arrears rates (fail to reject H0)",
    ),
)
p_value_arrears_sixty_days_df

test  larger_p_value  smaller_p_value  \
0                     All users with top-up            0.68             0.32   
1                      Users that logged-in            0.65             0.35   
2   Viewed inforcard (vs. login in control)            0.10             0.90   
3  Clicked inforcard (vs. login in control)            0.02             0.98   

                                                                                          result  
0                                         Samples have similar arrears rates (fail to reject H0)  
1                                         Samples have similar arrears rates (fail to reject H0)  
2                                         Samples have similar arrears rates (fail to reject H0)  
3  Variant arrears rates are significantly higher the control (reject H0 but undesirable effect)

**Conclusion:** Here we generally have no significant decrease in the arrears rates. We have an undesirable increase of 1 percentage point for variant users who clicked on the infocars vs. login users in control.

<a id='section2.2'></a>
## 6 month write-offs

In [38]:
sixty_day_write_off_query = """
with totals as (
select 
case when pu.user_in_control_group then 'control' else 'variant' end as user_split,
case when has_write_off = 1 then 'Has Write-off' else 'No Write-off' end as write_off_type,
count(*) as n_users,
count(case when has_login = 1 then 1 end) as login_users, 
count(case when n_views>0 then 1 end) as infocard_view_users,
count(case when n_action_clicked>0 then 1 end) as infocard_clicked_users
from sixty_day_df pu
left join sixty_day_infocards_df using (user_id)
where created <= '2021-06-30'
group by 1, 2
)
select *, 
round(n_users::numeric/ (sum(n_users) over (partition by user_split))::numeric, 3)*100 as perc_all_users,
round(login_users::numeric/ (sum(login_users) over (partition by user_split))::numeric, 3)*100 as perc_login_users,
round(infocard_view_users::numeric/ (sum(infocard_view_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_view_users,
round(infocard_clicked_users::numeric/ (sum(infocard_clicked_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_click_users
from totals
order by 2, 1
"""

In [39]:
sixty_day_write_off_df = con.execute(sixty_day_write_off_query).fetchdf()
sixty_day_write_off_df

user_split write_off_type  n_users  login_users  infocard_view_users  \
0    control  Has Write-off       93           90                    0   
1    variant  Has Write-off       87           86                   59   
2    control   No Write-off    21951        20423                    0   
3    variant   No Write-off    22000        20461                16205   

   infocard_clicked_users  perc_all_users  perc_login_users  \
0                       0             0.4               0.4   
1                       8             0.4               0.4   
2                       0            99.6              99.6   
3                    1767            99.6              99.6   

   perc_infocard_view_users  perc_infocard_click_users  
0                       NaN                        NaN  
1                       0.4                        0.5  
2                       NaN                        NaN  
3                      99.6                       99.5

In [40]:
experiment_charts(
    sixty_day_write_off_df[sixty_day_write_off_df["write_off_type"] == "Has Write-off"],
    "write_offs",
    1,
    color_scale_1,
)

alt.HConcatChart(...)

In [41]:
write_off_sixty_all_larger = z_test(
    sixty_day_user_data("sixty_day_variant_users_with_write_off"),
    sixty_day_user_data("sixty_day_variant_users"),
    sixty_day_user_data("sixty_day_control_users_with_write_off"),
    sixty_day_user_data("sixty_day_control_users"),
    "larger",
)
write_off_sixty_login_larger = z_test(
    sixty_day_user_data("sixty_day_variant_users_with_write_off_and_login"),
    sixty_day_user_data("sixty_day_variant_users_with_login"),
    sixty_day_user_data("sixty_day_control_users_with_write_off_and_login"),
    sixty_day_user_data("sixty_day_control_users_with_login"),
    "larger",
)
write_off_sixty_view_larger = z_test(
    sixty_day_user_data("sixty_day_variant_users_with_write_off_and_view"),
    sixty_day_user_data("sixty_day_variant_users_with_view"),
    sixty_day_user_data("sixty_day_control_users_with_write_off_and_login"),
    sixty_day_user_data("sixty_day_control_users_with_login"),
    "larger",
)
write_off_sixty_click_larger = z_test(
    sixty_day_user_data("sixty_day_variant_users_with_write_off_and_click"),
    sixty_day_user_data("sixty_day_variant_users_with_click"),
    sixty_day_user_data("sixty_day_control_users_with_write_off_and_login"),
    sixty_day_user_data("sixty_day_control_users_with_login"),
    "larger",
)

p_value_write_off_sixty_days_df = pd.DataFrame(
    {
        "test": [
            "All users with top-up",
            "Users that logged-in",
            "Viewed inforcard (vs. login in control)",
            "Clicked inforcard (vs. login in control)",
        ],
        "larger_p_value": [
            write_off_sixty_all_larger,
            write_off_sixty_login_larger,
            write_off_sixty_view_larger,
            write_off_sixty_click_larger,
        ],
    }
)

p_value_write_off_sixty_days_df["smaller_p_value"] = (
    1 - p_value_write_off_sixty_days_df["larger_p_value"]
)
p_value_write_off_sixty_days_df["result"] = np.where(
    p_value_write_off_sixty_days_df["smaller_p_value"] <= 0.05,
    "Variant write_off rates are significantly lower than the control (reject H0)",
    np.where(
        p_value_write_off_sixty_days_df["smaller_p_value"] >= 0.95,
        "Variant write_off rates are significantly higher the control (reject H0 but undesirable effect)",
        "Samples have similar write_off rates (fail to reject H0)",
    ),
)
p_value_write_off_sixty_days_df

test  larger_p_value  smaller_p_value  \
0                     All users with top-up            0.57             0.43   
1                      Users that logged-in            0.46             0.54   
2   Viewed inforcard (vs. login in control)            0.78             0.22   
3  Clicked inforcard (vs. login in control)            0.18             0.82   

                                                     result  
0  Samples have similar write_off rates (fail to reject H0)  
1  Samples have similar write_off rates (fail to reject H0)  
2  Samples have similar write_off rates (fail to reject H0)  
3  Samples have similar write_off rates (fail to reject H0)

**Conclusion:** Here we have no significant decrease in the write-off rates.

<a id='section3'></a>
# Fifty Percent usage infocard test results

<a id='section3.1'></a>
## 3 month arrears

In [42]:
fifty_perc_query = """
select * from experiment_df
where experiment_name = 'fifty-percent-usage-exceeded'
"""

fifty_perc_sample_query = """
select 
case when user_in_control_group then 'control' else replace(experiment_outcome, 'infocard_overdraft_limit_', '') end as user_split,
count(*) as n_users
from fifty_perc_df
group by 1
"""

In [43]:
fifty_perc_df = con.execute(fifty_perc_query).fetchdf()

con.register("fifty_perc_df", fifty_perc_df)
fifty_perc_sample_df = con.execute(fifty_perc_sample_query).fetchdf()

In [44]:
fifty_perc_arrears_query = """
with totals as (
select 
case when pu.user_in_control_group then 'control' else replace(pu.experiment_outcome, 'infocard_overdraft_limit_', '') end as user_split,
case when has_arrears = 1 then 'Has arrears' else 'No arrears' end as arrears_type,
count(*) as n_users,
count(case when has_login = 1 then 1 end) as login_users, 
count(case when n_views>0 then 1 end) as infocard_view_users,
count(case when n_action_clicked>0 then 1 end) as infocard_clicked_users
from fifty_perc_df pu
left join fifty_perc_infocards_df using (user_id)
group by 1, 2
)
select *, 
round(n_users::numeric/ (sum(n_users) over (partition by user_split))::numeric, 3)*100 as perc_all_users,
round(login_users::numeric/ (sum(login_users) over (partition by user_split))::numeric, 3)*100 as perc_login_users,
round(infocard_view_users::numeric/ (sum(infocard_view_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_view_users,
round(infocard_clicked_users::numeric/ (sum(infocard_clicked_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_click_users
from totals
order by 2, 1
"""

In [45]:
fifty_perc_arrears_df = con.execute(fifty_perc_arrears_query).fetchdf()

In [46]:
fifty_perc_arrears_df

user_split arrears_type  n_users  login_users  infocard_view_users  \
0           control  Has arrears     5658         4781                    0   
1  reminder_percent  Has arrears     2676         2310                 2434   
2      reminder_use  Has arrears     2765         2333                 2513   
3           control   No arrears    42051        36741                    0   
4  reminder_percent   No arrears    21148        18443                17780   
5      reminder_use   No arrears    21051        18360                17681   

   infocard_clicked_users  perc_all_users  perc_login_users  \
0                       0            11.9              11.5   
1                     646            11.2              11.1   
2                     740            11.6              11.3   
3                       0            88.1              88.5   
4                    3689            88.8              88.9   
5                    3957            88.4              88.7   

   perc_infocard_view_users  perc_infocard_click_users  
0                       NaN                        NaN  
1                      12.0                       14.9  
2                      12.4                       15.8  
3                       NaN                        NaN  
4                      88.0                       85.1  
5                      87.6                       84.2

In [47]:
# Z-Test Function
def fifty_perc_user_data(col):
    return fifty_perc_test_df.iloc[0][col].astype(float)


fifty_perc_test_query = """
select 
--All Fifty Percent Users
sum(case when sd.user_in_control_group then 1 end) as fifty_perc_control_users,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' then 1 end) as fifty_perc_variant_use_users,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' then 1 end) as fifty_perc_variant_perc_users,
sum(case when sd.user_in_control_group and has_top_up = 1 then 1 end) as fifty_perc_control_users_with_top_up,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and has_top_up = 1 then 1 end) as fifty_perc_variant_use_users_with_top_up,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and has_top_up = 1 then 1 end) as fifty_perc_variant_perc_users_with_top_up,
sum(case when sd.user_in_control_group and has_arrears = 1 then 1 end) as fifty_perc_control_users_with_arrears,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and has_arrears = 1 then 1 end) as fifty_perc_variant_use_users_with_arrears,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and has_arrears = 1 then 1 end) as fifty_perc_variant_perc_users_with_arrears,
sum(case when sd.user_in_control_group and has_write_off = 1 then 1 end) as fifty_perc_control_users_with_write_off,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and has_write_off = 1 then 1 end) as fifty_perc_variant_use_users_with_write_off,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and has_write_off = 1 then 1 end) as fifty_perc_variant_perc_users_with_write_off,
-- Fifty Percent Users With Login
sum(case when sd.user_in_control_group and has_login = 1 then 1 end) as fifty_perc_control_users_with_login,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and has_login = 1 then 1 end) as fifty_perc_variant_use_users_with_login,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and has_login = 1 then 1 end) as fifty_perc_variant_perc_users_with_login,
sum(case when sd.user_in_control_group and has_login = 1 and has_top_up = 1 then 1 end) as fifty_perc_control_users_with_top_up_and_login,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and has_login = 1 and has_top_up = 1 then 1 end) as fifty_perc_variant_use_users_with_top_up_and_login,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and has_login = 1 and has_top_up = 1 then 1 end) as fifty_perc_variant_perc_users_with_top_up_and_login,
sum(case when sd.user_in_control_group and has_login = 1 and has_arrears = 1 then 1 end) as fifty_perc_control_users_with_arrears_and_login,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and has_login = 1 and has_arrears = 1 then 1 end) as fifty_perc_variant_use_users_with_arrears_and_login,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and has_login = 1 and has_arrears = 1 then 1 end) as fifty_perc_variant_perc_users_with_arrears_and_login,
sum(case when sd.user_in_control_group and has_login = 1 and has_write_off = 1 then 1 end) as fifty_perc_control_users_with_write_off_and_login,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and has_login = 1 and has_write_off = 1 then 1 end) as fifty_perc_variant_use_users_with_write_off_and_login,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and has_login = 1 and has_write_off = 1 then 1 end) as fifty_perc_variant_perc_users_with_write_off_and_login,
-- Fifty Percent Users With Infocard Views
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and n_views >0 then 1 end) as fifty_perc_variant_use_users_with_view,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and n_views >0 then 1 end) as fifty_perc_variant_perc_users_with_view,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and n_views >0 and has_top_up = 1 then 1 end) as fifty_perc_variant_use_users_with_top_up_and_view,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and n_views >0 and has_top_up = 1 then 1 end) as fifty_perc_variant_perc_users_with_top_up_and_view,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and n_views >0 and has_arrears = 1 then 1 end) as fifty_perc_variant_use_users_with_arrears_and_view,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and n_views >0 and has_arrears = 1 then 1 end) as fifty_perc_variant_perc_users_with_arrears_and_view,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and n_views >0 and has_write_off = 1 then 1 end) as fifty_perc_variant_use_users_with_write_off_and_view,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and n_views >0 and has_write_off = 1 then 1 end) as fifty_perc_variant_perc_users_with_write_off_and_view,
-- Fifty Percent Users With Infocard Clicks
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and n_action_clicked >0 then 1 end) as fifty_perc_variant_use_users_with_click,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and n_action_clicked >0 then 1 end) as fifty_perc_variant_perc_users_with_click,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and n_action_clicked >0 and has_top_up = 1 then 1 end) as fifty_perc_variant_use_users_with_top_up_and_click,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and n_action_clicked >0 and has_top_up = 1 then 1 end) as fifty_perc_variant_perc_users_with_top_up_and_click,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and n_action_clicked >0 and has_arrears = 1 then 1 end) as fifty_perc_variant_use_users_with_arrears_and_click,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and n_action_clicked >0 and has_arrears = 1 then 1 end) as fifty_perc_variant_perc_users_with_arrears_and_click,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_use' and n_action_clicked >0 and has_write_off = 1 then 1 end) as fifty_perc_variant_use_users_with_write_off_and_click,
sum(case when sd.experiment_outcome = 'infocard_overdraft_limit_reminder_percent' and n_action_clicked >0 and has_write_off = 1 then 1 end) as fifty_perc_variant_perc_users_with_write_off_and_click
from fifty_perc_df sd
left join fifty_perc_infocards_df using (user_id)
order by 2, 1
"""
fifty_perc_test_df = con.execute(fifty_perc_test_query).fetchdf()

In [48]:
experiment_charts(
    fifty_perc_arrears_df[fifty_perc_arrears_df["arrears_type"] == "Has arrears"],
    "arrears",
    20,
    color_scale_2,
)

alt.HConcatChart(...)

#### Arrears Z-Test results for the Percent Reminder Infocard

In [49]:
# Reminder percent
arrears_fifty_perc_all_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_arrears"),
    fifty_perc_user_data("fifty_perc_variant_perc_users"),
    fifty_perc_user_data("fifty_perc_control_users_with_arrears"),
    fifty_perc_user_data("fifty_perc_control_users"),
    "larger",
)
arrears_fifty_perc_login_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_arrears_and_login"),
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_arrears_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)
arrears_fifty_perc_view_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_arrears_and_view"),
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_view"),
    fifty_perc_user_data("fifty_perc_control_users_with_arrears_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)
arrears_fifty_perc_click_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_arrears_and_click"),
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_click"),
    fifty_perc_user_data("fifty_perc_control_users_with_arrears_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)

p_value_arrears_fifty_perc_df = pd.DataFrame(
    {
        "test": [
            "All users with top-up",
            "Users that logged-in",
            "Viewed inforcard (vs. login in control)",
            "Clicked inforcard (vs. login in control)",
        ],
        "larger_p_value": [
            arrears_fifty_perc_all_larger,
            arrears_fifty_perc_login_larger,
            arrears_fifty_perc_view_larger,
            arrears_fifty_perc_click_larger,
        ],
    }
)

p_value_arrears_fifty_perc_df["smaller_p_value"] = (
    1 - p_value_arrears_fifty_perc_df["larger_p_value"]
)

p_value_arrears_fifty_perc_df["result"] = np.where(
    p_value_arrears_fifty_perc_df["smaller_p_value"] <= 0.05,
    "Variant arrears rates are significantly lower than the control (reject H0)",
    np.where(
        p_value_arrears_fifty_perc_df["smaller_p_value"] >= 0.95,
        "Variant arrears rates are significantly higher the control (reject H0 but undesirable effect)",
        "Samples have similar arrears rates (fail to reject H0)",
    ),
)
p_value_arrears_fifty_perc_df

test  larger_p_value  smaller_p_value  \
0                     All users with top-up            0.99             0.01   
1                      Users that logged-in            0.92             0.08   
2   Viewed inforcard (vs. login in control)            0.03             0.97   
3  Clicked inforcard (vs. login in control)            0.00             1.00   

                                                                                          result  
0                     Variant arrears rates are significantly lower than the control (reject H0)  
1                                         Samples have similar arrears rates (fail to reject H0)  
2  Variant arrears rates are significantly higher the control (reject H0 but undesirable effect)  
3  Variant arrears rates are significantly higher the control (reject H0 but undesirable effect)

**Conclusion:** Here we have somewhat contradictory results, with a significant decrease for all variant users of 0.7 percentage points, no significant changes for users that logged-in, and undesirable significantly higher arrears rates for the variant groups who either viewed or clicked the infocard (with increases of 0.5 percentage points and 3.4 percentage points respectively).

#### Arrears Z-Test results for the Usage Reminder Infocard

In [50]:
# Reminder usage
arrears_fifty_usage_all_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_use_users_with_arrears"),
    fifty_perc_user_data("fifty_perc_variant_use_users"),
    fifty_perc_user_data("fifty_perc_control_users_with_arrears"),
    fifty_perc_user_data("fifty_perc_control_users"),
    "larger",
)
arrears_fifty_usage_login_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_use_users_with_arrears_and_login"),
    fifty_perc_user_data("fifty_perc_variant_use_users_with_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_arrears_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)
arrears_fifty_usage_view_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_use_users_with_arrears_and_view"),
    fifty_perc_user_data("fifty_perc_variant_use_users_with_view"),
    fifty_perc_user_data("fifty_perc_control_users_with_arrears_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)
arrears_fifty_usage_click_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_use_users_with_arrears_and_click"),
    fifty_perc_user_data("fifty_perc_variant_use_users_with_click"),
    fifty_perc_user_data("fifty_perc_control_users_with_arrears_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)

p_value_arrears_fifty_usage_df = pd.DataFrame(
    {
        "test": [
            "All users with top-up",
            "Users that logged-in",
            "Viewed inforcard (vs. login in control)",
            "Clicked inforcard (vs. login in control)",
        ],
        "larger_p_value": [
            arrears_fifty_usage_all_larger,
            arrears_fifty_usage_login_larger,
            arrears_fifty_usage_view_larger,
            arrears_fifty_usage_click_larger,
        ],
    }
)

p_value_arrears_fifty_usage_df["smaller_p_value"] = (
    1 - p_value_arrears_fifty_usage_df["larger_p_value"]
)

p_value_arrears_fifty_usage_df["result"] = np.where(
    p_value_arrears_fifty_usage_df["smaller_p_value"] <= 0.05,
    "Variant arrears rates are significantly lower than the control (reject H0)",
    np.where(
        p_value_arrears_fifty_usage_df["smaller_p_value"] >= 0.95,
        "Variant arrears rates are significantly higher the control (reject H0 but undesirable effect)",
        "Samples have similar top-up rates (fail to reject H0)",
    ),
)
p_value_arrears_fifty_usage_df

test  larger_p_value  smaller_p_value  \
0                     All users with top-up            0.84             0.16   
1                      Users that logged-in            0.81             0.19   
2   Viewed inforcard (vs. login in control)            0.00             1.00   
3  Clicked inforcard (vs. login in control)            0.00             1.00   

                                                                                          result  
0                                          Samples have similar top-up rates (fail to reject H0)  
1                                          Samples have similar top-up rates (fail to reject H0)  
2  Variant arrears rates are significantly higher the control (reject H0 but undesirable effect)  
3  Variant arrears rates are significantly higher the control (reject H0 but undesirable effect)

**Conclusion:** And here we find no significant differences between variant and control in the all users and users who logged in groups, but we do have undesirable significantly higher arrears rates for the variant groups who either viewed or clicked the infocard (with increases of 0.9 percentage points and 4.3 percentage points respectively).

<a id='section3.2'></a>
## 6 month write-offs

In [51]:
fifty_perc_write_off_query = """
with totals as (
select 
case when pu.user_in_control_group then 'control' else replace(pu.experiment_outcome, 'infocard_overdraft_limit_', '') end as user_split,
case when has_write_off = 1 then 'Has write-off' else 'No write-off' end as write_off_type,
count(*) as n_users,
count(case when has_login = 1 then 1 end) as login_users, 
count(case when n_views>0 then 1 end) as infocard_view_users,
count(case when n_action_clicked>0 then 1 end) as infocard_clicked_users
from fifty_perc_df pu
left join fifty_perc_infocards_df using (user_id)
where created <= '2021-06-30'
group by 1, 2
)
select *, 
round(n_users::numeric/ (sum(n_users) over (partition by user_split))::numeric, 3)*100 as perc_all_users,
round(login_users::numeric/ (sum(login_users) over (partition by user_split))::numeric, 3)*100 as perc_login_users,
round(infocard_view_users::numeric/ (sum(infocard_view_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_view_users,
round(infocard_clicked_users::numeric/ (sum(infocard_clicked_users) over (partition by user_split))::numeric, 3)*100 as perc_infocard_click_users
from totals
order by 2, 1
"""

In [52]:
fifty_perc_write_off_df = con.execute(fifty_perc_write_off_query).fetchdf()

In [53]:
fifty_perc_write_off_df

user_split write_off_type  n_users  login_users  infocard_view_users  \
0           control  Has write-off      512          345                    0   
1  reminder_percent  Has write-off      250          185                  185   
2      reminder_use  Has write-off      235          156                  157   
3           control   No write-off    38624        33787                    0   
4  reminder_percent   No write-off    19240        16822                16767   
5      reminder_use   No write-off    19274        16840                16789   

   infocard_clicked_users  perc_all_users  perc_login_users  \
0                       0             1.3               1.0   
1                      38             1.3               1.1   
2                      32             1.2               0.9   
3                       0            98.7              99.0   
4                    3754            98.7              98.9   
5                    4052            98.8              99.1   

   perc_infocard_view_users  perc_infocard_click_users  
0                       NaN                        NaN  
1                       1.1                        1.0  
2                       0.9                        0.8  
3                       NaN                        NaN  
4                      98.9                       99.0  
5                      99.1                       99.2

In [54]:
experiment_charts(
    fifty_perc_write_off_df[
        fifty_perc_write_off_df["write_off_type"] == "Has write-off"
    ],
    "write-offs",
    2,
    color_scale_2,
)

alt.HConcatChart(...)

#### Write-off Z-Test results for the Percent Reminder Infocard

In [55]:
# Reminder percent
write_off_fifty_perc_all_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_write_off"),
    fifty_perc_user_data("fifty_perc_variant_perc_users"),
    fifty_perc_user_data("fifty_perc_control_users_with_write_off"),
    fifty_perc_user_data("fifty_perc_control_users"),
    "larger",
)
write_off_fifty_perc_login_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_write_off_and_login"),
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_write_off_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)
write_off_fifty_perc_view_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_write_off_and_view"),
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_view"),
    fifty_perc_user_data("fifty_perc_control_users_with_write_off_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)
write_off_fifty_perc_click_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_write_off_and_click"),
    fifty_perc_user_data("fifty_perc_variant_perc_users_with_click"),
    fifty_perc_user_data("fifty_perc_control_users_with_write_off_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)

p_value_write_off_fifty_perc_df = pd.DataFrame(
    {
        "test": [
            "All users with top-up",
            "Users that logged-in",
            "Viewed inforcard (vs. login in control)",
            "Clicked inforcard (vs. login in control)",
        ],
        "larger_p_value": [
            write_off_fifty_perc_all_larger,
            write_off_fifty_perc_login_larger,
            write_off_fifty_perc_view_larger,
            write_off_fifty_perc_click_larger,
        ],
    }
)

p_value_write_off_fifty_perc_df["smaller_p_value"] = (
    1 - p_value_write_off_fifty_perc_df["larger_p_value"]
)

p_value_write_off_fifty_perc_df["result"] = np.where(
    p_value_write_off_fifty_perc_df["smaller_p_value"] <= 0.05,
    "Variant write_off rates are significantly lower than the control (reject H0)",
    np.where(
        p_value_write_off_fifty_perc_df["smaller_p_value"] >= 0.95,
        "Variant write_off rates are significantly higher the control (reject H0 but undesirable effect)",
        "Samples have similar write_off rates (fail to reject H0)",
    ),
)
p_value_write_off_fifty_perc_df

test  larger_p_value  smaller_p_value  \
0                     All users with top-up            0.63             0.37   
1                      Users that logged-in            0.17             0.83   
2   Viewed inforcard (vs. login in control)            0.14             0.86   
3  Clicked inforcard (vs. login in control)            0.45             0.55   

                                                     result  
0  Samples have similar write_off rates (fail to reject H0)  
1  Samples have similar write_off rates (fail to reject H0)  
2  Samples have similar write_off rates (fail to reject H0)  
3  Samples have similar write_off rates (fail to reject H0)

**Conclusion:** Here we have no significant decrease in the write-off rates.

#### Write-off Z-Test results for the Usage Reminder Infocard

In [56]:
# Reminder usage
write_off_fifty_usage_all_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_use_users_with_write_off"),
    fifty_perc_user_data("fifty_perc_variant_use_users"),
    fifty_perc_user_data("fifty_perc_control_users_with_write_off"),
    fifty_perc_user_data("fifty_perc_control_users"),
    "larger",
)
write_off_fifty_usage_login_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_use_users_with_write_off_and_login"),
    fifty_perc_user_data("fifty_perc_variant_use_users_with_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_write_off_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)
write_off_fifty_usage_view_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_use_users_with_write_off_and_view"),
    fifty_perc_user_data("fifty_perc_variant_use_users_with_view"),
    fifty_perc_user_data("fifty_perc_control_users_with_write_off_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)
write_off_fifty_usage_click_larger = z_test(
    fifty_perc_user_data("fifty_perc_variant_use_users_with_write_off_and_click"),
    fifty_perc_user_data("fifty_perc_variant_use_users_with_click"),
    fifty_perc_user_data("fifty_perc_control_users_with_write_off_and_login"),
    fifty_perc_user_data("fifty_perc_control_users_with_login"),
    "larger",
)

p_value_write_off_fifty_usage_df = pd.DataFrame(
    {
        "test": [
            "All users with top-up",
            "Users that logged-in",
            "Viewed inforcard (vs. login in control)",
            "Clicked inforcard (vs. login in control)",
        ],
        "larger_p_value": [
            write_off_fifty_usage_all_larger,
            write_off_fifty_usage_login_larger,
            write_off_fifty_usage_view_larger,
            write_off_fifty_usage_click_larger,
        ],
    }
)

p_value_write_off_fifty_usage_df["smaller_p_value"] = (
    1 - p_value_write_off_fifty_usage_df["larger_p_value"]
)

p_value_write_off_fifty_usage_df["result"] = np.where(
    p_value_write_off_fifty_usage_df["smaller_p_value"] <= 0.05,
    "Variant write_off rates are significantly lower than the control (reject H0)",
    np.where(
        p_value_write_off_fifty_usage_df["smaller_p_value"] >= 0.95,
        "Variant write_off rates are significantly higher the control (reject H0 but undesirable effect)",
        "Samples have similar write_off rates (fail to reject H0)",
    ),
)
p_value_write_off_fifty_usage_df

test  larger_p_value  smaller_p_value  \
0                     All users with top-up            0.88             0.12   
1                      Users that logged-in            0.85             0.15   
2   Viewed inforcard (vs. login in control)            0.80             0.20   
3  Clicked inforcard (vs. login in control)            0.83             0.17   

                                                     result  
0  Samples have similar write_off rates (fail to reject H0)  
1  Samples have similar write_off rates (fail to reject H0)  
2  Samples have similar write_off rates (fail to reject H0)  
3  Samples have similar write_off rates (fail to reject H0)

**Conclusion:** Here we have no significant decrease in the write-off rates.